Add replay profile command to GAPIT

apbodnar opened the pull request:
This renames the current `profile` command to `coarse_profile`.  `profile` now takes a perfetto trace of a replay using a provided config.  It will use a simple render stages only config if no config is provided.

RenfengLiu commented on cmd/gapit/profile.go:
```diff
@@ -16,31 +16,53 @@ package main
 
-type profileVerb struct{ GetTimestampsFlags }
+const (
+	bufferSizeKb                            = uint32(131072)
```
can we have some comments about those numbers?

apbodnar replied:
Sure, I shamelessly took them from gapis/trace/android/trace.go  It's possible that we should use different values in any case.

RenfengLiu commented on gapis/replay/gpu_profile.go:
```diff
@@ -0,0 +1,57 @@
+		return nil, err
+	}
+
+	if device != nil {
```
if we're checking `device != nil ` then maybe we also need to deal with the `else` condition.

pmuetschard commented on gapir/client/session.go:
```diff
@@ -327,6 +328,8 @@ func (s *session) newADB(ctx context.Context, d adb.Device, abi *device.ABI, lau
 
 	log.I(ctx, "Launching GAPIR...")
+	// Configure GAPIR to be traceable for replay profiling
+	perfetto_android.SetupRenderStagesEnvironment(ctx, d, apk.Name, true, abi, []string{})
```
Don't ignore the error or the cleanup here.

pmuetschard commented on gapis/service/service.proto:
```diff
@@ -1287,6 +1290,16 @@ message TraceTypeCapabilities {
+message GpuProfileResponse {
+  string trace_save_path = 1;
+}
+
```
This is not the API we've agreed upon and looks to me like a step in the wrong direction. If you need to save the trace to disk to debug things, add a `config.go` option that will dump it somewhere, similar to the "dump replay" stuff.

pmuetschard commented on cmd/gapit/profile.go:
```diff
@@ -16,27 +16,33 @@ package main
 
+	"github.com/google/gapid/core/os/file"
+	"github.com/google/gapid/gapis/config"
+	perfetto_pb "perfetto/config"
```
This one should probably be in it's own group after all the gapid imports.

pmuetschard commented on gapis/perfetto/android/trace.go:
```diff
@@ -71,8 +71,8 @@ func setupVulkanLayers(ctx context.Context, d adb.Device, packageName string, ab
 
-func setupRenderStagesEnvironment(ctx context.Context, d adb.Device, packageName string, perfettoConfig *perfetto_pb.TraceConfig, abi *device.ABI, layers []string) (app.Cleanup, error) {
-	if !hasRenderStageEnabled(perfettoConfig) {
+func SetupRenderStagesEnvironment(ctx context.Context, d adb.Device, packageName string, hasRenderStages bool, abi *device.ABI, layers []string) (app.Cleanup, error) {
```
Since you are making this function public, please rename it. It's a bit silly that it says "do X" and takes a boolean that controls whether "do do X"... Maybe `SetupProfileLayers`

Also, since it's now public, it needs a doc comment.

pmuetschard commented on gapis/service/service.proto:
```diff
@@ -1287,6 +1290,15 @@ message TraceTypeCapabilities {
+message GpuProfileRequest {
+  path.Capture capture = 1;
+  path.Device device = 2;
+  TraceOptions trace_options = 3;
```
Don't pass the trace options here. That is way too generic. I think the only thing we need here are the capture and the device.

Another reason why passing the perfetto config from the client doesn't work here: you'd have to setup all the layers (not just the GPU slices layer) when starting gapir.

I think all you have to do is move the code that creates the trace options from gapit into gapis.

AWoloszyn replied:
We shouldn't have to do much for the layers, we could put "all" of the layers in the right place when starting GAPIR always, then we can turn them on/off at instance creation time I think.

pmuetschard replied:
The system setting overrides the `VkInstanceCreateInfo`. I don't think you can "turn-off" any layers the system requests.

AWoloszyn replied:
But we can **not** add these layers to the system setting, and just add them to the VkInstanceCreateInfo (in a transform).

apbodnar commented:
@AWoloszyn @pmuetschard I addressed your feedback.  Instead of returning a local path to the trace, it now returns a (currently) empty `ProfilingData` proto with the option of dumping the trace via config.go

purvisa-at-google-com commented on gapis/replay/gpu_profile.go:
```diff
@@ -0,0 +1,88 @@
+
+// Eyeball some generous trace config parameters
+const (
+	bufferSizeKb                            = uint32(262144)
```
256 * 1024 is clearer to read. The compiler will constant fold to 262144 for you.

AWoloszyn commented on gapir/client/session.go:
```diff
@@ -327,6 +328,17 @@ func (s *session) newADB(ctx context.Context, d adb.Device, abi *device.ABI, lau
 
 	log.I(ctx, "Launching GAPIR...")
+	// Configure GAPIR to be traceable for replay profiling
+	cleanup, err := perfetto_android.SetupProfileLayers(ctx, d, apk.Name, true, abi, []string{})
```
This makes me nervous. This means no matter WHAT we are doing the profiling layers will be turned on in GAPIR. That seems like something we do not want. I do not trust that the profiling layers will be .. stable enough .. to handle all of the odd stuff we do in GAPID replay that ISN'T just "replay this frame".

hevrard replied:
Would the solution be to add the profiling layers using a transform like the FindIssue one which adds the validation layers?
See https://github.com/google/gapid/blob/1e1473feeb407daa5ddfa7491b0dc1032b70a8d9/gapis/api/vulkan/find_issues.go#L85

apbodnar replied:
I've added an additional commit that does the layer loading explicitly. It still requires some config, but at least gapir will not load the stages layer when not in a replay.

hevrard commented on cmd/gapit/coarse_profile.go:
```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2018 Google Inc.
```
copyright 2019

hevrard commented on gapis/trace/android/adreno/profiling_data.go:
```diff
@@ -0,0 +1,12 @@
+package adreno
```
Missing copyright header

pmuetschard commented on gapis/trace/android/trace.go:
```diff
@@ -118,6 +118,25 @@ func (t *androidTracer) GetDevice() bind.Device {
+	if err != nil {
+		return nil, log.Err(ctx, err, "Failed to read trace buffer")
+	}
+	processor, err := perfetto.NewProcessor(ctx, rawData)
```
This leaks the processor. It needs to be closed.

pmuetschard commented on gapis/server/server.go:
```diff
@@ -914,6 +914,13 @@ func (s *server) GetTimestamps(ctx context.Context, req *service.GetTimestampsRe
 	return replay.GetTimestamps(ctx, req.Capture, req.Device, req.LoopCount, h)
 }
 
+func (s *server) GpuProfile(ctx context.Context, req *service.GpuProfileRequest) (*service.GpuProfileResponse, error) {
```
This should return `(service.ProfilingData, error)` instead.

pmuetschard replied:
See https://github.com/pmuetschard/gapid/commit/e8909ff5093f12ff5d25912c55aa6883d0bad6e2

pmuetschard commented on gapis/api/vulkan/replay.go:
```diff
@@ -781,13 +783,15 @@ type profileRequest struct {
 	device *device.Instance,
 	out transform.Writer) error {
 	transforms := transform.Transforms{}
+	transforms.Add(&profilingLayers{})
```
This transform should only be added if we're profiling **and** if the layer is available.

apbodnar replied:
Oops, actually didn't mean to leave it in.  For now, replay profiling will not use split replays

pmuetschard commented on gapis/api/vulkan/profiling_layers.go:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	switch cmd := cmd.(type) {
+	case *VkCreateInstance:
```
This will cause the replay to fail if the layer is not available. I would prefer if we either don't do anything if the layer is not available. At least add a TODO for now.

pmuetschard commented on gapis/trace/android/trace.go:
```diff
@@ -118,6 +118,26 @@ func (t *androidTracer) GetDevice() bind.Device {
+	gpuName := gpu.GetName()
+	if strings.Contains(gpuName, "Adreno") {
+		return adreno.ProcessProfilingData(ctx, processor)
+	}
```
I'm not certain this really needs to be GPU specific, but I'm OK to keep it in for now.

pmuetschard commented on gapis/replay/gpu_profile.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// GpuProfile replays the trace and writes a Perfetto trace of the replay
+func GpuProfile(ctx context.Context, capturePath *path.Capture, device *path.Device) (*service.GpuProfileResponse, error) {
```
This should return `(*service.ProfilingData, error)`. You are currently wrapping things, then unwrapping them in `server.go` just to wrap them again in `grpc.go` :) I know this is because of the change I've asked for before.
